File: 124.cpp
#include <iostream>
#include <vector>

using namespace std;
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
    int maxSum = INT_MIN;
    int pathSum(TreeNode* root) {
        if (root==nullptr) return 0;

        int r = pathSum(root->right);
        int l = pathSum(root->left);
        
        int currentMaxSum = l > 0 ? root->val + l + (r > 0 ? r : 0) : root->val + (r > 0 ? r : 0);
        if (currentMaxSum > maxSum) maxSum = currentMaxSum;
        
        return root->val + max(0, max(l, r)); //只能选左右子树中的一个或者不选
    }
public:
    int maxPathSum(TreeNode* root) {
        if (root==nullptr) return 0;
        pathSum(root);
        return maxSum;
    }
};

File: 124.go
package leetcode

import "math"

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func maxPathSum(root *TreeNode) int {
	ans := math.MinInt

	var dfs func(root *TreeNode) int
	dfs = func(root *TreeNode) int {
		if root == nil {
			return 0
		}

		maxSon := 0
		left := dfs(root.Left)
		maxSon = max(maxSon, left)

		right := dfs(root.Right)
		maxSon = max(maxSon, right)

		cur := root.Val
		// ans = max(ans, maxSon) //不需要考虑这种情况，因为子节点已经考虑过了
		ans = max(ans, cur+maxSon)
		ans = max(ans, cur+left+right)

		return cur + maxSon
	}

	dfs(root)
	return ans
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
